### PR TITLE
Disabled upgrades of modules if the installed version is higher than this proposed by Addons

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Disabled upgrades of modules if the installed version is higher than this proposed by Addons. We remove the upgrade of modules files in `upgradeFiles`, and now, we upgrade modules in `upgradeModules`.
| Type?             | new feature
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes PrestaShop/PrestaShop#14784
| How to test?      | Autoupgrade Prestashop from an old version to the last one for example. Most of native modules must be updated.

### CODE REVIEW

**Review by PierreRambaud — DISMISSED** [review]

File: classes/UpgradeTools/ModuleAdapter.php [hunk 1]
```diff
@@ -112,17 +113,37 @@ public function listModulesToUpgrade(array $modulesFromAddons)
         }
 
         foreach (scandir($dir) as $module_name) {
-            if (is_file($dir . DIRECTORY_SEPARATOR . $module_name)) {
+            // We don't update autoupgrade module
+            if ($module_name !== 'autoupgrade') {
```
Comment (atomiix) [inline classes/UpgradeTools/ModuleAdapter.php:117]: Am I misunderstanding something?
```suggestion
            if ($module_name === 'autoupgrade') {
```

Comment (Progi1984) [reply classes/UpgradeTools/ModuleAdapter.php]: Fixed; Thanks

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Progi1984** [comment]
Rebased after the @okom3pom PR

**Review by matthieu-rolland — APPROVED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by AureRita — CHANGES_REQUESTED** [review]
Hi @Progi1984 ! 

I tested it and it correctly keep the right version when you've a new version and put  others on the good version, ( tested with module brandlist 1.0.3 [https://github.com/PrestaShop/ps_brandlist/pull/12], the current version on develop was 1.0.2 ) , but this bring with it an exception on BO > Design > Link List 

as you can see on this picture : 

Exception on BO > Design > Link List
image

some little detail : 
- you've to do a hard refresh ( CTRL + Maj + R ) else you gonna have some UI problems
- I updated the version 1.7.7.8 to 1.7.8.4

**Comment by Progi1984** [comment]
@AureRita What was the version of `ps_linklist` before and after the migration ? I'm interested. I think the bug come from `ps_linklist`.

**Review by AureRita — CHANGES_REQUESTED** [review]
@Progi1984 

the version of ps_linklist before the migration was 4.0.0
image

and after it's still the same version ( it didn't update )

File: classes/TaskRunner/Upgrade/UpgradeModules.php [hunk 2]
```diff
@@ -56,7 +56,6 @@ public function run()
             return true;
         }
 
-        $time_elapsed = time() - $start_time;
```
Comment (matks) [inline classes/TaskRunner/Upgrade/UpgradeModules.php:59]: 👍

File: classes/UpgradeTools/ModuleAdapter.php [hunk 3]
```diff
@@ -129,17 +130,37 @@ public function listModulesToUpgrade(array $modulesFromAddons)
...
 
         foreach (scandir($dir) as $module_name) {
-            if (is_file($dir . DIRECTORY_SEPARATOR . $module_name)) {
+            // We don't update autoupgrade module
+            if ($module_name === 'autoupgrade') {
                 continue;
             }
-
+            // We have a file modules/mymodule
+            if (is_file($dir . $module_name)) {
```
Comment (matks) [inline classes/UpgradeTools/ModuleAdapter.php:138]: No `DIRECTORY_SEPARATOR` needed (previous code used it)?

**Review by matks — COMMENTED** [review]
Little question ;)

File: classes/UpgradeTools/ModuleAdapter.php (hunk 3 above)
Comment (Progi1984) [reply classes/UpgradeTools/ModuleAdapter.php:138]: As defined in `classes/UpgradeContainer.php`, it already contains a `DIRECTORY_SEPARATOR` at the end.

**Review by atomiix — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Review by AureRita — CHANGES_REQUESTED** [review]
Hi @Progi1984, 

I tested your correction and you keep the newest version of brand list but : 

In FO : 

We can't add some product on wishlist 
image
I think new module isn't add when you do a realese 

--------------

In BO : 

Orders > Invoices > Invoice Option  
The design of the page is the same that older 
image
It's the same on : Orders > Invoices > Delivery slips 
Catalogs > Categories 
Catalogs > Brand & Suppliers 
Customers > Customers 
Customers > Customers Service 
Customers > Merchandise Returns 
Design > Pages 
International > Localization > languages 
International > Localization > Currencies 
International > Localization > Geolocalisation 
International > Location > Zones
International > Location > Countries 
International > Taxes > Taxes
Shop Parameters > General > General 
Shop Parameters > General > Maintenance 
Shop Parameters > Order Settings > Order Stettings
Shop Parameters > Product Settings 
Shop Parameters > Customer Settings > Customer 
Shop Parameters > Traffics & SEO > Refferers 
Shop Parameters > Search > Search 
Advenced Parameters > Performance
Advenced Parameters > Administration 
Advenced Parameters > Emails 
Advenced Parameters > Import 
Advenced Parameters > Teams > Employees 
Advenced Parameters > Database > DB Backup
Advenced Parameters > Webservice 
Advenced Parameters > Experimental Features
 
-------------------

At Shop Parameters > Customer Settings > Groups, you've the old save button : 
image

---------------

You keep the succès button : 
image

-----------

On Shop Parameters > Contact > Contact, you've a button  " Recommended Modules and Services"
image

--------------

You've another subpart of "Shop Parameters" named Merchant Expertise, it can be related to the succes maybe : 
image

-----------

Bullet point with the old version on Advenced Parameters > Mails 
image

----------------

You've the older version of alert messages on Advenced Parameters > Import :
image

also on : 
Advenced Parameters > Database > SQL Manager 
Advenced Parameters > Database > DB Backup
Advenced Parameters > Webservice 
Advenced Parameters > Experimental Features

Thank You @Progi1984, good luck and if you need more info I can help you

**Comment by Progi1984** [comment]
@AureRita 
- New modules are not added.
- For the problem of design, I think it's not relative to this PR. I have already heard of this bug by @friends-of-presta community. 
- The success button is linked to `gamification` module. The button "Recommended Modules and Services" is linked to `ps_mbo` module. Which versions of the module do you have before the upgrade ?
- For design problems, I really don't think it's related to this PR (I work only on modules). Could you check it on `dev` branch of the module ?

**Comment by AureRita** [comment]
- Ok 
- Ok, I tested with the normal autoupdate and yes, you've all the issue related to design 
- MBO version before and after : 2.0.1 / Gamification version before and after ; 2.3.2
- Strangely, we have maybe an issue with the dev version with Prestashop 1.7.7.8 : 
image

**Comment by Progi1984** [comment]
@AureRita So it's QA ☑️ ?

**Comment by AureRita** [comment]
Yes, it is :)

**Comment by atomiix** [comment]
Thanks @Progi1984 & @AureRita

**Pull request merged**